lang/java/reef-runtime-multi/src/main/java/org/apache/reef/runtime/multi/driver/RuntimesHost.java

```diff
@@ -0,0 +1,169 @@
+
+import org.apache.avro.io.DecoderFactory;
+import org.apache.avro.io.JsonDecoder;
+import org.apache.avro.specific.SpecificDatumReader;
+import org.apache.reef.runtime.common.driver.api.*;
+import org.apache.reef.runtime.common.driver.resourcemanager.NodeDescriptorEvent;
+import org.apache.reef.runtime.common.driver.resourcemanager.ResourceAllocationEvent;
+import org.apache.reef.runtime.common.driver.resourcemanager.ResourceStatusEvent;
+import org.apache.reef.runtime.common.driver.resourcemanager.RuntimeStatusEvent;
+import org.apache.reef.runtime.multi.client.parameters.SerializedRuntimeDefinitions;
+import org.apache.reef.runtime.multi.driver.org.apache.reef.runtime.multi.driver.parameters.RuntimeName;
+import org.apache.reef.runtime.multi.utils.RuntimeDefinition;
+import org.apache.reef.tang.Configuration;
+import org.apache.reef.tang.Injector;
+import org.apache.reef.tang.JavaConfigurationBuilder;
+import org.apache.reef.tang.Tang;
+import org.apache.reef.tang.annotations.Parameter;
+import org.apache.reef.tang.exceptions.InjectionException;
+import org.apache.reef.tang.formats.AvroConfigurationSerializer;
+import org.apache.reef.wake.EventHandler;
+import org.apache.reef.wake.time.runtime.event.RuntimeStart;
+import org.apache.reef.wake.time.runtime.event.RuntimeStop;
+
+import javax.inject.Inject;
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * Hosts the actual runtime implementations and delegates invocations to them.
+ */
+final class RuntimesHost {
+  private final Set<String> runtimeDefinitions;
+  private Map<String, Runtime> runtimes;
+  private final Injector originalInjector;
+  private String defaultRuntimeName;
+
+  @Inject
+  private RuntimesHost(final Injector injector,
+                       @Parameter(SerializedRuntimeDefinitions.class) final Set<String> runtimeDefinitions) {
+    if (runtimeDefinitions == null || runtimeDefinitions.size() == 0) {
+      throw new RuntimeException("No runtime configurations are provided for multi-runtime");
+    }
+
+    this.runtimeDefinitions = runtimeDefinitions;
+    this.originalInjector = injector;
+  }
+
+  private static RuntimeDefinition parseSerializedRuntimeDefinition(final String serializedRuntimeDefinition) throws
+          IOException {
+    final RuntimeDefinition rd;
+    final JsonDecoder decoder = DecoderFactory.get().
+            jsonDecoder(RuntimeDefinition.getClassSchema(), serializedRuntimeDefinition);
+    final SpecificDatumReader<RuntimeDefinition> reader = new SpecificDatumReader<>(RuntimeDefinition.class);
+    rd = reader.read(null, decoder);
+    return rd;
+  }
+
+  private synchronized void initialize() {
+    if (this.runtimes != null) {
+      return;
+    }
+
+    this.runtimes = new HashMap<>();
+    final AvroConfigurationSerializer serializer = new AvroConfigurationSerializer();
+
+    for (final String serializedRuntimeDefinition : runtimeDefinitions) {
```

Comment:
Wouldn't it be easier to have a schema for the group of runtimes in addition to each runtime on its own? That way, you can add the `Configuration` instances directly and can also take care of the unique default runtime at the schema level instead of requiring code to check for it.